[WNMGDS-2115] Sass conversion to use CSS Custom Properties

forrestbaer opened the pull request:
## Summary

- Updated all core/hc/mc/docsite/storybook's scss to utilize css vars instead of sass variables, only sass variables to remain are in the `_layout.scss` files in each system
- Updated build system & tokens in many places to accomidate this change.
- Now copying theme files to `.storybook/static` and `docs/static` for use in storybook/doc site
- Updated doc site method for changing theme variables
- Updated storybook method for changing theme variables
- Updated snapshots
    - Some snapshots I was unable to fullly understand why they were different, the differences in some cases seem to have resolved some visual things that I personally think looked wrong to begin with, such as the distance between header and content in the modals and the size of the spinner icon in the spinner icon buttons. In those cases, the appropriate selector appears to be being used where it wasn't previously.. Anyway will need to go over these a bit more to confirm with designers.
- Several small visual bugfixes along the way
- Adds a new script for building storybook `scripts/build-storybook.sh` which moves static assets appropriately when building storybook for the docs/gatsby site.

## How to test

1. `yarn clean && yarn install && yarn build && yarn build:healthcare && yarn build:medicare`
2. `yarn build:storybook && yarn build:storybook:docs && yarn build:docs && yarn serve:docs`
3. `yarn storybook` `npx http-server ./storybook-static`
4. Evaluate snapshot updates.
5. Evaluate css selector sequencing in dev tools for snapshot updated components next. to `design.cms.gov/storybook` versions.
6. `yarn test`

## Checklist

- [x] Prefixed the PR title with the [Jira ticket number](https://jira.cms.gov/projects/WNMGDS/) as `[WNMGDS-####] Title`

pwolfert commented on packages/design-system-tokens/src/copy_themes.sh:
```diff
@@ -13,16 +13,15 @@ copyThemes()
-  done
-  for file in *.css; do
-    cp -v "$file" "../../docs/src/styles/theme-variables/_${file}"
+    rm "${PATHS[$i]}/${i}-theme.css"
+    echo ":root, ::before, ::after, ::backdrop {" >> "${PATHS[$i]}/${i}-theme.css"
+    cat "${i}-theme.css" >> "${PATHS[$i]}/${i}-theme.css"
+    cat "${i}-components-theme.css" >> "${PATHS[$i]}/${i}-theme.css"
+    echo "}" >> "${PATHS[$i]}/${i}-theme.css"
+    # copy to storybook static for storybook building
+    cp -v "${PATHS[$i]}/${i}-theme.css" ../../../.storybook/static
+    mkdir -p ../../docs/static/themes
+    cp -v "${PATHS[$i]}/${i}-theme.css" ../../docs/static/themes
```
It might be easier to read if `"${PATHS[$i]}/${i}-theme.css"` were turned into a local variable. It looks like it's being used six times, so it could be justified.

forrestbaer replied:
> I haven't dug into why, but the little arrow icons aren't showing for me on the dropdowns. I just stumbled upon that randomly. Might have to do with the image path variable.

Hmm that's odd, they're showing up for me.

![image](https://user-images.githubusercontent.com/783820/208765740-ab41d918-1ca4-4746-8a3a-bf4a0d60ea66.png)

pwolfert replied:
I cleaned and built again, and it wasn't a problem anymore. False alaram

pwolfert commented on packages/design-system-tokens/src/themes/core-components.ts:
```diff
@@ -249,6 +249,7 @@ export const components: AnyTokenValues = {
     '__border-color--checked':                    t.color['primary'],
     '__border-color--disabled':                   t.color['gray-light'],
     '__border-color--error':                      t.color['base'], // Don't use error border for choice in this theme
+    '__border-color--error--inverse':             t.color['white'],
```
👍

pwolfert commented on packages/design-system-tokens/src/themes/medicare-components.ts:
```diff
@@ -348,7 +349,7 @@ export const components: AnyTokenValues = {
   },
   
   'review': {
-    'review__border-color':                       t.color['border'],
+    '__border-color':                       t.color['border'],
```
Nice catch

pwolfert commented on packages/docs/src/components/layout/Layout.tsx:
```diff
@@ -68,6 +68,12 @@ const Layout = ({
+          rel="stylesheet"
+          type="text/css"
+          title="docThemeCss"
+          href={`/themes/${theme}-theme.css`}
```
Oh hey, the fact that this is declarative makes this nice and simple

pwolfert commented on packages/docs/src/styles/index.scss:
```diff
@@ -1,29 +1,30 @@
-@import 'theme-variables/_medicare-theme.css';
-@import 'theme-variables/_medicare-components-scss-to-css.map.scss';
-@import 'theme-variables/_medicare-components-theme.css';
+// // importing core styles from core index.scss files
+// @import 'theme-variables/_core-scss-to-css.map.scss';
+// @import 'theme-variables/_core-theme.css';
+// @import 'theme-variables/_core-components-scss-to-css.map.scss';
+// @import 'theme-variables/_core-components-theme.css';
+// // brand themes will use core variables if the theme does not explicitly define that variable
+// @import 'theme-variables/_healthcare-scss-to-css.map.scss';
+// @import 'theme-variables/_healthcare-theme.css';
+// @import 'theme-variables/_healthcare-components-scss-to-css.map.scss';
+// @import 'theme-variables/_healthcare-components-theme.css';
+// @import 'theme-variables/_medicare-scss-to-css.map.scss';
+// @import 'theme-variables/_medicare-theme.css';
+// @import 'theme-variables/_medicare-components-scss-to-css.map.scss';
+// @import 'theme-variables/_medicare-components-theme.css';
```
Can these lines be removed now?

pwolfert commented on packages/docs/src/styles/index.scss:
```diff
@@ -1,29 +1,30 @@
-  // Have to set the image path so that the error label's svg can be found
-  $image-path: '@cmsgov/ds-healthcare-gov/src/images';
-  @import '@cmsgov/ds-healthcare-gov/src/styles/components/index';
-}
+// [data-theme='healthcare'] {
+//   // Have to set the image path so that the error label's svg can be found
+//   $image-path: '@cmsgov/ds-healthcare-gov/src/images';
+//   @import '@cmsgov/ds-healthcare-gov/src/styles/index';
+// }
```
Checking my understanding: We don't need to worry about including healthcare- or medicare-specific CSS in the doc site CSS because we don't directly embed any of those components and instead rely on the storybook styles being complete, right?

forrestbaer replied:
Just the fonts, I will look closer at fonts in this regard tomorrow.

pwolfert commented:
I can see how the spinner buttons were not following the intention of 

```scss
.ds-c-button > & {
    @include inline-icon;
    @include spinner-size(1em, 1em);
  }
```

because of the old namespaced theme rules having greater specificity! I'm glad this is fixed!

<img width="715" alt="Screen Shot 2022-12-20 at 11 37 33 AM" src="https://user-images.githubusercontent.com/7595652/208753300-34a3bd4f-95f3-4b6b-a700-d01fc2f5c30a.png">

pwolfert reviewed (commented):
I'm not finished going through this, but here are my notes so far. This was a big project!

pwolfert commented:
<img width="481" alt="Screen Shot 2022-12-20 at 11 52 28 AM" src="https://user-images.githubusercontent.com/7595652/208754747-28f2936a-21aa-465d-b910-5a2ca96eb734.png">

I haven't dug into why, but the little arrow icons aren't showing for me on the dropdowns. I just stumbled upon that randomly. Might have to do with the image path variable.

forrestbaer commented:
I added a new script `scripts/build-storybook.sh` and set up the npm scripts to use it. This script moves the `images` and `fonts` folders when building storybook for the docs site into the `docs/static/` folder so they can be referenced by storybook stories and gatsby in a central location.

I've also updated the aliases for gatsby `images` and `fonts` to use these static assets for building. This seems to resolve an issue I was running into where storybook stories were trying to reference one place and gatsby itself was looking elsewhere, I believe this is the most straightforward way to manage this issue with static assets without needing to have multiple copies of them all over the place.

zarahzachz commented on packages/design-system/src/styles/components/_TooltipIcon.scss:
```diff
@@ -40,13 +42,13 @@
-    border-color: $tooltip-icon__color;
-    border-color: rgba($tooltip-icon__color, 0.25);
+    border-color: var(--tooltip-icon__color);
+    border-color: rgba(var(--tooltip-icon__color), 0.25);
```
this color function shouldn't work with css vars - there's goofiness in how the compiled var value fits in with this function. 

sass allows for these color functions to work with sass vars, css props don't.

zarahzachz replied:
[SO explains it better than me](https://stackoverflow.com/a/41265350)

forrestbaer replied:
Interesting, and too bad! I've updated these variables to use new token variables which are computed during compilation.

zarahzachz commented on packages/design-system/src/styles/components/_TooltipIcon.scss:
```diff
@@ -56,8 +58,8 @@
-      border-color: $tooltip-icon__color--inverse;
-      border-color: rgba($tooltip-icon__color--inverse, 0.25);
+      border-color: var(--tooltip-icon__color--inverse);
+      border-color: rgba(var(--tooltip-icon__color--inverse), 0.25);
```
see "issues with css vars and color function" comment above

zarahzachz commented on packages/design-system/src/styles/components/_UsaBanner.scss:
```diff
@@ -14,7 +15,7 @@
   display: flex;
   flex-direction: row;
   font-size: 12px;
```
weird this wasn't a var

forrestbaer replied:
We don't have a font value for 12px in tokens.

zarahzachz commented on packages/design-system/src/styles/utilities/_grid.scss:
```diff
@@ -9,13 +10,13 @@
-  padding-left: calc(#{$grid-gutter-width} * 0.5);
-  padding-right: calc(#{$grid-gutter-width} * 0.5);
+  max-width: var(--site-max-width);
+  padding-left: calc(var(--grid-gutter-width) * 0.5);
```
the unitless division here seems to work! css is a mystery sometimes

zarahzachz reviewed (commented):
<img width="897" alt="Screen Shot 2022-12-22 at 5 13 59 PM" src="https://user-images.githubusercontent.com/5159392/209234771-77a72aed-ed24-4fc0-a0e5-a7bedee794be.png">
the border seems to have disappeared for the review pattern but i can't pinpoint why 🤔

zarahzachz commented:
i don't know if this is helpful or not, but for some of the css/scss stuff i reviewed, i ran the css solution through [codepen](https://codepen.io/your-work) (and used the computed panel in chrome dev tools on the demo) and the original sass implementation through [sassmeister](https://www.sassmeister.com/) just to compare what the outcomes were. this was how i spotted some discrepancies in `calc()` and color functions

forrestbaer commented:
I believe all those issues are resolved @zarahzachz , thanks for the review, good callouts!

pwolfert commented on scripts/build-storybook.sh:
```diff
@@ -0,0 +1,21 @@
+  mkdir -p packages/docs/static/images && \
+  cp -R packages/docs/static/storybook/fonts/* packages/docs/static/fonts/ && \
+  cp -R packages/docs/static/storybook/images/* packages/docs/static/images/ && \
+  rm -rf packages/docs/static/storybook/fonts && \
+  rm -rf packages/docs/static/storybook/images
```
Why do we remove these directories? Or why don't we need them?

forrestbaer replied:
I wanted to make sure the content in them was only what we want to copy over.

pwolfert commented on scripts/build-storybook.sh:
```diff
@@ -0,0 +1,21 @@
+  echo "${GREEN}info${NC}: Moving static assets for use with gatsby."
+  mkdir -p packages/docs/static/fonts && \
+  mkdir -p packages/docs/static/images && \
+  cp -R packages/docs/static/storybook/fonts/* packages/docs/static/fonts/ && \
+  cp -R packages/docs/static/storybook/images/* packages/docs/static/images/ && \
```
Should we consider ignoring these folders in git so we don't get so many duplicate files in version control?

pwolfert replied:
We ignore the `static/storybook` folder already

forrestbaer replied:
I suppose, will just need to make sure to not ignore them if any updates are made there.

pwolfert reviewed (commented):
I'm noticing some fairly substantial spacing changes in the medicare footer. Do you know what's causing that?

forrestbaer commented:
@pwolfert looks like an `@extend` was causing some trouble there, I'm replacing that extend with the appropriate vars now and it's working. Updating snapshots.

pwolfert reviewed (approved):
Now that they've been committed, I think you now need to explicitly remove those image and font files from git even though they're being ignored for future changes. Everything else LGTM 🎉
